Build recent classifications component of signed-in home page

goplayoutside3 opened the pull request:
## Package
lib-user
app-root

## Linked Issue and/or Talk Post
Homepages Project Board: https://github.com/orgs/zooniverse/projects/36/views/2

## Describe your changes
- Created a UserHome component in lib-user. This component renders in app-root when a signed-in user visits the homepage
- Created a RecentSubjects component in lib-user. This is similar to RecentSubjects on a project classify page, but not project-specific
- Created a SubjectCard component. This card intentionally does **not** match the Figma design for the signed-in homepage. I've doublechecked with Sean in-office, and decided to create a SubjectCard similar to those in app-project for now. Eventually there will be a shared SubjectCard used throughout FEM on Recents pages, Collections pages, Favorites pages, Classify pages, and the signed-in home page, and development of it does not need to happen now.

## How to Review

- [ ]  Run lib-user's Storybook and view stories for RecentSubjects and SubjectCard
- [ ] Run app-root locally with `yarn build` and `yarn start` to view https://local.zooniverse.org:3000 while signed-in. You'll see the Layout from lib-user, but the internal layout of UserHome will be updated in a future PRs

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)

shaunanoordin commented on packages/lib-user/src/components/UserHome/components/RecentSubjects/RecentSubjects.js:
```diff
@@ -0,0 +1,68 @@
+        style={{ listStyle: 'none' }}
+        margin='0'
+      >
+        {isLoading && (
+          <Box fill justify='center' align='center'>
+            <Loader />
+          </Box>
+        )}
```
[ui] Loaders = good! 👍 I always appreciate status feedback for users.

[ui][future improvement] We should also figure out a UI standard for showing error messages, should something go sideways.

goplayoutside3 replied:
Good point! I'll add a basic error message to this component for now.

shaunanoordin commented on packages/lib-user/src/components/UserHome/components/RecentSubjects/RecentSubjects.js:
```diff
@@ -0,0 +1,68 @@
+            </Text>
+          </Box>
+        )}
+        {!isLoading && subjects?.length
+          ? subjects
+              .slice(0, 10)
+              .map(subject => <SubjectCard key={subject.id} size={size} subject={subject} />)
+          : null}
```
[super minor][coding] This uses the ternary so we have `!isLoading && subjects?.length ? subjects.slice().map(...) : null`, but wouldn't the rendered effects still be the same if we write it as `!isLoading && subjects?.slice().map(...)` ? I thought running a render-mapping an empty array was valid - i.e. `[].map(x => <div>{x}</div)` returns `[]`, which React renders as nothing but doesn't break.

Actually, you know what? Shaun TODO: double check this. A quick search for "React render empty array" on the Goog shows differing opinions on the interwebs.

shaunanoordin commented on packages/lib-user/src/components/UserHome/components/RecentSubjects/RecentSubjectsContainer.js:
```diff
@@ -0,0 +1,37 @@
+    userId: authUser.id
+  }
+  const {
+    data: subjects,
```
⚠️ The data being fetches by fetchRecentSubjects() is actually **classification resources,** not **subject resources.**

For example, "subjects" here might actually look something like...

```
[
  {
    id: "258337208",  // Classification ID
    links: {
      project: "19072",
      workflow: "22152",
      subject: "80286011"  // Actual Subject ID
    },
    ...
  },
...
]
```

eatyourgreens replied:
That’s not a Classification, it’s a Recent, which is a Panoptes resource representing your recent subjects.

eatyourgreens replied:
There’s some code in the project app that uses these for the Classify page.

goplayoutside3 replied:
Correct - are you suggesting code changes in this file?

eatyourgreens replied:
I don’t think so. Just clarifying that the ID isn’t a classification ID or a subject ID.

eatyourgreens replied:
Oh, I see, you call them `classification` in the code. The code works, but the name might be confusing if someone expects them to be Classification objects.

In the project app, they're named `recent`. It's a tiny thing, but might make the code easier to follow.

https://github.com/zooniverse/front-end-monorepo/blob/4deea7e0476c7352e768da6eaadfed1ee0ccf169/packages/app-project/stores/User/Recents/Recents.js#L48-L54

goplayoutside3 replied:
Gotcha, thanks!

shaunanoordin reviewed (approved):
## PR Review

Package: app-root, lib-user

This PR adds some content to the Signed-In Homepage.

- On the signed-in homepage:
  - (new feature) Recent Subjects: displays recently classified Subjects of the current signed-in user.
    - Shows up to 10 Subjects. Clicking on a Subject should take the user to the Subject's page on the project's Talk.
      - ❔ Only tested with Image-type subjects, so who knows how this will look like for Video-types or Data-types or multi-types. 🤔 
      - [low priority] Limitation: links are only expected to work with Production subjects, not Staging subjects.  
      - ❗ Bug: clicking on a Subject DOESN'T take users to the Subject's page on Talk as expected. See Dev Notes below.
    - Has a loading status indicator, and a message when user has 0 Subjects. 👍 
    - No error status indicator, however.

_Screenshot: user darkeshard's Recent Classifications, as seen on the signed-in home page on app-root_

<img width="662" alt="image" src="https://github.com/zooniverse/front-end-monorepo/assets/13952701/0c29b3ba-979c-4a26-b671-fcfb6b02ecaa">

### Dev Notes

Problem: clicking on a Subject DOESN'T take users to the Subject's page on Talk as expected.

- If you take a look at the screenshot above, the first (left-most) Subject is listed as Subject 258337208.
- Clicking on that Subject leads to https://www.zooniverse.org/projects/spidercrabwatch/spider-crab-watch/talk/subjects/258337208 , which shows no valid Subject

Analysis:
- This is because this ISN'T Subject 258337208.
- It's more accurately **Classification** 258337208, linked to Subject **80286011.**
- The correct link should to the Subject's page on Talk should be https://www.zooniverse.org/projects/spidercrabwatch/spider-crab-watch/talk/subjects/80286011 
- I think the confusion here stems from `RecentSubjectsContainer` and/or `fetchRecentSubjects()` - the data being fetched is _classification resources,_ not _subject resources._

Recommended Solution:
- Most of the code should work as is, we just need to 1. clarify when we're fetching Classifications vs Subjects, and 2. make minor tweaks to RecentSubjects and/or SubjectCard
- e.g. `fetchRecentSubjects()` => `fetchRecentClassifications()`
- and then changing `<SubjectCard>` to something like...

```
export default function SubjectCard({ size = 'medium', classification }) {
  const subjectMedia = classification.locations.map(
    location => Object.values(location)[0]
  )
  const mediaSrc = subjectMedia[0]

  // to PFE
  const href = `https://www.zooniverse.org/projects/${classification.slug}/talk/subjects/${classification.links.subject.id}`

  ...
}
```

Huh, come to think of it, it's just that `href` that needs to be fixed to accommodate the "correct view" of classifications instead of subjects.

### Testing

Tested on app-root on localhost, with macOS + Chrome 125

Testing steps:
- Open the home page, set for production: https://local.zooniverse.org:3000/?env=production
- Log in as a user who has previously made classifications (e.g. darkeshard)
- Observe that recently classified Subjects are displayed. Confirm that Subject IDs are correct.
- Click on a Subject to be taken to the Subject's page on the project's Talk.

### Status

This PR looks mostly good to me, save for that one small and easily fixed bug. I'm approving this PR, but requesting that issue be fixed before you hit "merge".

Bugs to be fixed before merging:
- [ ] Clicking on a Subject should link to the correct page on Talk. i.e. SubjectCards should use the correct Subject ID, not Classification ID.

goplayoutside3 commented:
I refactored the props passed to SubjectCard and grabbed the correct `subjectID` for the href. To answer your question about other types of subjects, the Media component from lib-react-components will handle all types of subjects `locations` including image, video, text, and json. I've added some more SubjectCard stories to this PR with examples of each.

I'll make a few more minor improvements to this PR before merge. Thanks for the review!

goplayoutside3 commented on packages/lib-user/src/components/UserHome/components/SubjectCard/SubjectCard.js:
```diff
@@ -0,0 +1,104 @@
+  const mediaSrc = subjectMedia[0]
+
+  // to PFE
+  const href = `https://www.zooniverse.org/projects/${subject.slug}/talk/subjects/${subject.id}`
```
True - I'd like to implement some better handling of `env` in app-root, and we might be able to apply `?env=staging` or `?env=production` when needed.
